gravitee-node-notifier/src/main/java/io/gravitee/node/notifier/trigger/NotificationTrigger.java

```diff
@@ -0,0 +1,230 @@
+import io.gravitee.node.api.notifier.NotificationAcknowledgeRepository;
+import io.gravitee.node.api.notifier.NotificationCondition;
+import io.gravitee.node.api.notifier.NotificationDefinition;
+import io.gravitee.node.notifier.NotifierUtils;
+import io.gravitee.node.notifier.plugin.NotifierPluginFactory;
+import io.gravitee.notifier.api.Notification;
+import io.gravitee.notifier.api.Notifier;
+import io.reactivex.Maybe;
+import io.vertx.core.Context;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.util.Date;
+import java.util.Optional;
+import java.util.Random;
+import java.util.function.Consumer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.scheduling.support.CronExpression;
+
+/**
+ * @author Eric LELEU (eric.leleu at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class NotificationTrigger implements Handler<Long> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(
+    NotificationTrigger.class
+  );
+
+  private Vertx vertx;
+
+  private NotificationAcknowledgeRepository notificationAcknowledgeRepository;
+
+  private NotifierPluginFactory notifierFactory;
+
+  private NotificationDefinition definition;
+
+  private NotificationCondition condition;
+
+  private Long scheduledTaskId;
+
+  private CronExpression cronExpression;
+
+  private final int randomDelayInSec;
+
+  public NotificationTrigger(
+    Vertx vertx,
+    NotificationAcknowledgeRepository notificationAcknowledgeRepository,
+    NotifierPluginFactory notifierFactory,
+    NotificationDefinition definition,
+    NotificationCondition condition,
+    boolean tryToAvoidMultipleNotif
+  ) {
+    this.vertx = vertx;
+    this.notificationAcknowledgeRepository = notificationAcknowledgeRepository;
+    this.notifierFactory = notifierFactory;
+    this.definition = definition;
+    this.condition = condition;
+    this.cronExpression = CronExpression.parse(definition.getCron());
+    if (tryToAvoidMultipleNotif) {
+      this.randomDelayInSec = Math.max(1, new Random().nextInt(10));
+    } else {
+      this.randomDelayInSec = -1;
+    }
+  }
+
+  public void start() {
+    scheduleNextAttempt();
+  }
+
+  private void scheduleNextAttempt() {
+    this.scheduledTaskId = this.vertx.setTimer(computeNextAttempt(), this);
+  }
+
+  public void stop() {
+    if (
+      this.scheduledTaskId != null &&
+      this.vertx.cancelTimer(this.scheduledTaskId)
+    ) {
+      LOGGER.debug("Notification Trigger cancelled !");
+    } else {
+      LOGGER.debug("Notification Trigger can't be cancelled or doesn't exist");
+    }
+  }
+
+  private long computeNextAttempt() {
+    final LocalDateTime now = LocalDateTime.now();
+    final long delay =
+      toEpochMillis(this.cronExpression.next(now)) - toEpochMillis(now);
+    return randomDelayInSec > 0 ? delay + randomDelayInSec : delay;
```

Comment:
Seems we mix millisecond and second here no ? 🧐

Reply:
indeed :mag: